new CLP: CollectSamErrorMetric

yfarjoun opened the pull request:
This contains a new CLP which collects base error information on a sam/bam file based on various covariates.

There are two error types (simple and overlapping bases) and many many covariates that can be combined in arbitrary ways

yfarjoun commented:
I'm writing (even) more unit tests..but I thought that we could get the bulk of the review started!

magicDGS commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+    public BaseErrorCalculation.Errors ERRORS_VALUE = null;
+
+    @Argument(doc = "A fake argument used to show what the options of STRATIFIER (in ERROR_METRICS) are.", optional = true)
+    public ReadBaseStratification.Stratifiers STRATIFIER_VALUE = null;
```
I believe that there is an easier way to handle this with barclay - maybe a plugin ([`CommandLinePluginDescriptor`](https://github.com/broadinstitute/barclay/blob/master/src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java)) or a tagged-argument ([`TaggedArgument`](https://github.com/broadinstitute/barclay/blob/master/src/main/java/org/broadinstitute/barclay/argparser/TaggedArgument.java))? Or a combination of both?

yfarjoun commented:
Perhaps but until the legacy parser is removed from Picard I think there
should be a way for those users to get information... Thanks for looking!

On Wed, May 30, 2018 at 4:20 AM Daniel Gómez-Sánchez <
notifications@github.com> wrote:

> *@magicDGS* commented on this pull request.
> ------------------------------
>
> In src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java
> <https://github.com/broadinstitute/picard/pull/1180#discussion_r191679833>
> :
>
> > +            "ERROR,READ_GROUP",
> +            "ERROR,MISMATCHES_IN_READ",
> +            "ERROR,ONE_BASE_PADDED_CONTEXT",
> +            "OVERLAPPING_ERROR",
> +            "OVERLAPPING_ERROR,BASE_QUALITY",
> +            "OVERLAPPING_ERROR,INSERT_LENGTH",
> +            "OVERLAPPING_ERROR,READ_ORDINALITY",
> +            "OVERLAPPING_ERROR,READ_ORDINALITY,CYCLE",
> +            "OVERLAPPING_ERROR,READ_ORDINALITY,HOMOPOLYMER",
> +            "OVERLAPPING_ERROR,READ_ORDINALITY,GC_CONTENT"));
> +
> +    @Argument(doc = "A fake argument used to show what the options of ERROR (in ERROR_METRICS) are.", optional = true)
> +    public BaseErrorCalculation.Errors ERRORS_VALUE = null;
> +
> +    @Argument(doc = "A fake argument used to show what the options of STRATIFIER (in ERROR_METRICS) are.", optional = true)
> +    public ReadBaseStratification.Stratifiers STRATIFIER_VALUE = null;
>
> I believe that there is an easier way to handle this with barclay - maybe
> a plugin (CommandLinePluginDescriptor
> <https://github.com/broadinstitute/barclay/blob/master/src/main/java/org/broadinstitute/barclay/argparser/CommandLinePluginDescriptor.java>)
> or a tagged-argument (TaggedArgument
> <https://github.com/broadinstitute/barclay/blob/master/src/main/java/org/broadinstitute/barclay/argparser/TaggedArgument.java>)?
> Or a combination of both?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/picard/pull/1180#pullrequestreview-124294458>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnk0v37IHdtKuoux91D_n_y1bnc9R2bks5t3lZTgaJpZM4USMwv>
> .
>

magicDGS commented:
True - I keep forgetting that picard is still using the legacy parser. Thanks for this work, looks great for analyzing some of my datasets!

lbergelson commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,203 @@
+        }
+
+        private boolean areReadsMates(final SAMRecord read1, final SAMRecord read2) {
+            // must have same name
```
should you be checking for secondary / supplementary reads here?

yfarjoun replied:
secondary ones sure....but supplementary I think are reasonable to leave.

lbergelson commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,203 @@
+     * A calculator that estimates the error rate of the bases it observes, assuming that the reference is truth.
+     * This calculator only includes bases that have been read twice in the same template (once from each read) and
+     * thus only includes bases that arise from the overlapping part of the reads. Over those bases the Calculator
+     * distinguishes between weather the two reads agree with each other but differ from the reference (indicative
```
typo: 
![image](https://user-images.githubusercontent.com/4700332/40743587-f5a63fa2-641f-11e8-8351-dff6d41e52e6.png)

yfarjoun replied:
it raining typos!

lbergelson commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,898 @@
+ */
+public class ReadBaseStratification {
+    // This variable has to be set using setLongHomopolymer _before_ the first call to binnedHomopolymerStratifier.get()
+    // if you need different binned Homopolimers with different LONG_HOMOPOLYMER values, you'll need to make your own
```
polimer

lbergelson commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+     * @param sequenceDictionary
+     * @return the intersection of the interval lists pointed to by the input parameter INTERVALS
+     */
+    private IntervalList getIntervals(final SAMSequenceDictionary sequenceDictionary) {
```
doesn't this exist somewhere else already?

yfarjoun replied:
not that I'm aware of...picard has very few tools that operate on intervals...

yfarjoun commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,147 @@
+        public long TOTAL_BASES;
+
+        public ComputableMetricBase(final String covariate, final long totalBases) {
+            this.TOTAL_BASES = totalBases;
```
All the metrics variables in picard are CAPITALS_WITH_UNDERSCORES...why complain now?

yfarjoun commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,147 @@
+        public int THREE_WAYS_DISAGREEMENT_Q = 0;
+
+        @Override
+        public void calculateDerivedFields() {
```
because these metrics can also be modified post-facto (not immutable) and they implement mergeable, which further allows them to be modified.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,215 @@
+            // one must be first while the other is not
+            if (!read1.getFirstOfPairFlag() ^ read2.getFirstOfPairFlag()) return false;
+            // one must be second while the other is not
+            if (!read1.getSecondOfPairFlag() ^ read2.getSecondOfPairFlag()) return false;
```
how about:
```
if ((read1.getFirstOfPairFlag() && read2. getSecondOfPairFlag()) || (read1.getSecondOfPairFlag() && read2. getFirstOfPairFlag())
```
You're not wrong, it just might link the fact that one should be r1 and the other r2.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "the sequencing experiment: position in the read, base context, length of insert and so on. " +
+                "" +
+                "This program enables the user to collect two different kinds of error metrics (one which attempts to " +
+                "distinguish between pre- and post- sequencer errors, and on which doesn't) and a collation of 'stratifiers' " +
```
`on` -> `one`

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "For example:\n" +
+                "The BASE_QUALITY stratifier will place bases in bins according to their declared base quality. " +
+                "The READ_ORDINALITY stratifier will place bases in one of two bins depending on whether their read " +
+                "is 'first' or 'second'. One could generate a composite stratifier BASE_QUALITY,READ_ORDINALITY which will " +
```
first or second _of a pair_

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "Sequencing errors come in different 'flavors', for example some occur during sequencing while " +
+                "others happen during 'library construction'. They can be correlated with various aspect of " +
+                "the sequencing experiment: position in the read, base context, length of insert and so on. " +
+                "" +
```
Missing newline `\n` in the empty strings here and below?  Can you build the tool and print it's usage to verify it looks good?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "do both as the same time. " +
+                "" +
+                "The resulting metric file will be named according to a provided prefix and a suffix which is generated " +
+                " automatically according to the error metric. " +
```
space before automatically

nh13 replied:
comma after prefix, to make it clear that one is provided by the user and the other is provided by the tool?

nh13 replied:
What are the extensions?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                " automatically according to the error metric. " +
+                "To save on run-time, the tool collects many metrics in a single pass and there should be hardly any " +
+                "performance loss when collecting many metrics at the same time, which is why the default already includes a " +
+                "lengthy collection of metrics. " +
```
`many` -> `multiple` here and above?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "lengthy collection of metrics. " +
+                "" +
+                "The tool takes in a VCF for the sample " +
+                "to skip actually polymorphic bases. As a result it assumes that any disagreement from the reference is an error." +
```
`actually polymorphic bases` -> `polymorphic sites`.  It doesn't do the "cute" thing and treat the alternate alleles also as reference (not errors), but simply skips those sites.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "the error rates that are exposed. " +
+                "" +
+                "The program also takes a region over which to run. This region should be used to define a 'confidence region' " +
+                "where lack of a call in the VCF translates to a high probability of a monomorphic site.",
```
I don't see the connection to the VCF as described.  The VCF is just used to skip sites that are assumed to have non-reference alleles, while the regions over which to run are simply that.  The latter region _could_ be generated from the target-capture regions, or creating a set of regions with high coverage, or looking at the gVCF, or just the NIST high-confidence regions, for a start, so I think it is confusing to have the second sentence (I was confused).

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                "The program also takes a region over which to run. This region should be used to define a 'confidence region' " +
+                "where lack of a call in the VCF translates to a high probability of a monomorphic site.",
+
+        oneLineSummary = "Program to collect error metrics on bases stratified in various ways.",
```
How about `collect base-level error metrics`?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+    @Argument(shortName = StandardOptionDefinitions.INPUT_SHORT_NAME, doc = "Input SAM or BAM file.")
+    public File INPUT;
+
+    @Argument(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "Basename for output files.")
```
CollectMultipleMetrics has _Base name_ (with a space).

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+    @Argument(shortName = "V", doc = "VCF of loci to avoid collecting data (should include any known variation!)")
+    public File VCF=null;
+
+    @Argument(shortName = "L", doc = "Region(s) to limit analysis to. Can be given as a interval_list file. (will intersect if multiple values are given.) ", optional = true)
```
> Can be given as a interval_list file

What else can it be given as?  Also `interval_list file` needs to be updated (see how other tools describe the `INTERVALS` option in Picard).

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+    @Argument(shortName = "L", doc = "Region(s) to limit analysis to. Can be given as a interval_list file. (will intersect if multiple values are given.) ", optional = true)
+    public List<File> INTERVALS = null;
+
+    @Argument(shortName = StandardOptionDefinitions.MINIMUM_MAPPING_QUALITY_SHORT_NAME, doc = "Minimum Mapping quality to include read.", minValue = 0)
```
Mapping -> mapping?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+    @Argument(shortName = StandardOptionDefinitions.MINIMUM_MAPPING_QUALITY_SHORT_NAME, doc = "Minimum Mapping quality to include read.", minValue = 0)
+    public int MIN_MAPPING_Q = 20;
+
+    @Argument(shortName = "BQ", doc = "Minimum Base quality to include base.",minValue = 0)
```
nit-picks: space after comma and Base -> base

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+            return computeQScore(numberOfErrors, TOTAL_BASES);
+        }
+
+
```
nitpick: extra newline (BINGO)

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+    @Argument(shortName = "PE", doc = "The prior error, in phred-scale (used for calculating ratios and filtering out HET sites)",
+            optional = true, minValue = 2)
+    public int PRIOR_Q = 30;
```
Without reading the code, I have no idea what the prior is used for.  The description the brackets don't help.  I tried reading the code and couldn't figure it out either :/

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFilesAreReadable(INTERVALS);
+
```
Want to make sure the output is writeable?  It should still be OK that it is a output prefix.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+            final SAMSequenceDictionary sequenceDictionary = referenceSequenceFileWalker.getSequenceDictionary();
+            if (sam.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate)
+                throw new PicardException("Input BAM must be sorted by coordinate");
```
brackets around `if` body please

nh13 replied:
Input could be a SAM or CRAM too!

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+ @CommandLineProgramProperties(
+        summary = "Program to collect error metrics on bases stratified in various ways." +
+                "\n" +
+                "Sequencing errors come in different 'flavors', for example some occur during sequencing while " +
```
Mention somewhere the input must be coordinate sorted.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+            if (sam.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate)
+                throw new PicardException("Input BAM must be sorted by coordinate");
+
+            sequenceDictionary.assertSameDictionary(sam.getFileHeader().getSequenceDictionary());
```
How about checking the VCF?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+            } else {
+                log.info("Intersecting interval_list: ", intervalListFile, ".");
+            }
+            regionOfInterest = IntervalList.intersection(regionOfInterest, temp);
```
uniqued? sorted?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+        for (final File intervalListFile : INTERVALS) {
+
+            final IntervalList temp;
+            if (intervalListFile.exists()) {
```
Wasn't IOUtil.assertFilesAreReadable run on the interval files already?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+            log.info("Using " + aggregatorList.size() + " aggregators.");
+
+            aggregatorList.forEach(la ->
+                    IOUtil.assertFileIsWritable(new File(OUTPUT + la.getSuffix())));
```
why the line wrap?  See line 235 for length.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+            // This hasNext() call has side-effects. It loads up the index and makes sure that the iterator is really ready for
+            // action. Calling this allows for the logging to be more accurate.
+            iterator.hasNext();
+            log.info("Really starting iteration now.");
```
How about changing this and the previous log message to be "preparing traversal" and "starting traversal" or something similar?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+            final IntervalList regionOfInterest = getIntervals(sequenceDictionary);
+
+            log.info("Getting SamLocusIterator");
```
Users wont know what a `SamLocusIterator` is.  How about `loading inputs`?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+            for (final SAMLocusAndReferenceIterator.SAMLocusAndReference info : iterator) {
+                if (random.nextDouble() > PROBABILITY)
+                    continue;
```
nitpick: brackets around `if` body.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+     * if vc and l are in the same contig the return value will be the number of bases apart they are,
+     * otherwise it will be MIN_INT/MAX_INT
+     */
+    public static int CompareVariantContextToLocus(final SAMSequenceDictionary dictionary, final VariantContext vc, final Locus l) {
```
CompareVariantContextToLocus -> compareVariantContextToLocus

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+        if (l.getPosition() < vc.getStart())
+            return vc.getStart() - l.getPosition();
+        if (l.getPosition() > vc.getEnd())
+            return vc.getEnd() - l.getPosition();
```
brackets?

nh13 replied:
Do you want to ignore deletions somewhere?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                progressLogger.record(info.locus.getSequenceName(), info.locus.getPosition());
+
+                if (MAX_LOCI != 0 && nProcessedLoci >= MAX_LOCI) {
+                    log.warn("Early stopping due to having processed MAX_LOCI loci.");
```
Stopping early?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+                }
+            }
+
+        } catch (IOException e) {
```
final?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+        long nProcessedLoci = 0;
+
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFilesAreReadable(INTERVALS);
```
Assert REFERENCE_SEQUENCE is readable?  How about the VCF? Then you don't need the ugly try/catch below.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+        } catch (IOException e) {
+            log.error("A problem occurred:", e.getMessage());
+            return 1;
```
remove the try/catch by asserting the various input file are readable?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+    }
+
+    /**
+     * Iterate over the different records in the locus and add bases to aggregators
```
nitpick: period at end.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+     * @return Collection of {@link BaseErrorAggregation}
+     * @throws IllegalArgumentException if directives cannot be parsed or objects cannot be created.
+     */
+
```
nitpick: extra newline, my favorite

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+    private Collection<BaseErrorAggregation> getAggregatorList() {
+        final List<BaseErrorAggregation> aggregatorList = new ArrayList<>();
+        Set<String> suffixes = new HashSet<>();
```
can this be final?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+
+    /**
+     * Reads Interprets intervals from the input INTERVALS, if there's a file with that name, it opens the file, otherwise it tries to parse it, checks that their dictionaries all agree with the input SequenceDictionary
+     * and returns the intersection of all the lists.
```
This sentence needs rewriting (ex. `Reads Interprets intervals `)

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+        final int numberOfDirectives = ParsingUtils.split(stratifierDirective, directiveUnits, ',', false);
+
+        if (numberOfDirectives == 256) {
+            throw new RuntimeException("Cannot parse more than 255 'directives'. Sorry. (What are you doing?)");
```
Why is this your business?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+        if (numberOfDirectives == 256) {
+            throw new RuntimeException("Cannot parse more than 255 'directives'. Sorry. (What are you doing?)");
+        }
+        if (numberOfDirectives == 0) {
```
nitpick: `else if`

nh13 commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,464 @@
+            throw new RuntimeException("Cannot parse more than 255 'directives'. Sorry. (What are you doing?)");
+        }
+        if (numberOfDirectives == 0) {
+            throw new RuntimeException("Found no directives at all. Cannot process.");
```
How about just `"No directives given"?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+        PRIOR_ERROR = priorError;
+    }
+
+    static double PRIOR_ERROR = 0.001;
```
doc?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+ */
+public class ErrorMetrics {
+    public static void setPriorError(double priorError) {
+        PRIOR_ERROR = priorError;
```
Why not just make PRIOR_ERROR public?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+
+    static double PRIOR_ERROR = 0.001;
+
+    public static class ErrorMetricBase extends MergeableMetricBase {
```
doc?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+     */
+    public static class OverlappingErrorMetric extends ErrorMetricBase {
+
+        /** The number of bases for which an overlapping base from the mate read was found*/
```
nitpick: space after found

nh13 replied:
How about `The number of bases that had a corresponding overlapping base from their mate`

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+        public OverlappingErrorMetric() {
+            super();
+        }
+
```
nitpick: new line

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+
+        /** The (phred) rate of bases that disagree with the reference */
+        @NoMergingIsDerived
+        public Integer Q_SCORE;
```
Why `Integer` versus `int`?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/Pair.java:
```diff
@@ -0,0 +1,112 @@
+ * @author Yossi Farjoun
+ */
+
+public class Pair<X extends Comparable<X>, Y extends Comparable<Y>> implements Comparable<Pair<X, Y>> {
```
See if you can extend `htsjdk.samtools.util.ComparableTuple`

nh13 commented on src/main/java/picard/sam/BamErrorMetric/Pair.java:
```diff
@@ -0,0 +1,112 @@
+        if (firstCompare != 0) return firstCompare;
+
+        if (this.second == null || o.second == null)
+            throw new NullPointerException("Found null objects in comparing " + toString() + " and " + o.toString());
```
nitpick: brackets around `if` body here and above

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,909 @@
+    }
+
+    /**
+     * defaults to 6
```
`defaults to 6` does not belong here, but perhaps at the class level.  The doc should tell us what this method does.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,909 @@
+    /**
+     * An enum designed to hold a binned version of
+     * any probability-like number (between 0 and 1)
+     * in quintiles
```
word wrap weirdness.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,909 @@
+        F1R2,
+        F2R1,
+        F1F2,
+        R1R2;
```
Why not use `PairOrientation` in htsjdk?

yfarjoun replied:
Pair orientation in htsJdk is Innie vs. Outtie. vs. Tandem, which is not what I have here.

nh13 replied:
See your doc comment, which _could_ be interpreted differently. Change `F1R2, F2R1, F1F2 or R1R2` to `F1R2, F2R1, F1F2, or R1R2`

nh13 commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,909 @@
+     * same order as was originally read) and the base under consideration is replaced with the reference base at that location
+     * All other bases are taken from the read, not the reference.
+     */
+
```
extra nl

nh13 commented on src/main/java/picard/sam/BamErrorMetric/SAMLocusAndReferenceIterator.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Iterator that traverses a SAM File and a ReferenceFile, accumulating information on a per-locus basis.
+ * Only, whatever loci are covered by the input reads are returned.
+ * Duplicate reads and non-primary alignments are filtered out.
```
This isn't controlled here, remove?

yfarjoun replied:
I"m going to try to move this to htsjdk anyhow...but sure.

nh13 commented on src/main/java/picard/sam/BamErrorMetric/SAMLocusAndReferenceIterator.java:
```diff
@@ -0,0 +1,81 @@
+            return locus.getRecordAndOffsets();
+        }
+    }
+}
```
I don't need it, but it could be nice to add to htsjdk.

yfarjoun replied:
I agree.

yfarjoun replied:
https://github.com/samtools/htsjdk/pull/1137

yfarjoun replied:
samtools/htsjdk#1137 was merged, once there's a htsjdk release, I'll upgrade to it and use that in my branch.

yfarjoun commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java:
```diff
@@ -0,0 +1,86 @@
+    public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SAMLocusAndReferenceIterator.SAMLocusAndReference locusInfo) {
+        final Object stratus = stratifier.stratify(recordAndOffset, locusInfo);
+        //this assumes we do not want to aggregate null.
+        if (stratus != null) strataAggregatorMap.get(stratus).addBase(recordAndOffset, locusInfo);
```
what would be the advantage of using Optional? The downside is an additional layer of boxing/unboxing

yfarjoun replied:
returning the stratus is not in the BaseCalculator interface...though I'm not sure that it's a good enough reason. What's the use-case for getting the stratus?

nh13 commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,215 @@
+        @Override
+        public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SAMLocusAndReferenceIterator.SAMLocusAndReference locusAndRef) {
+            final byte readBase = recordAndOffset.getReadBase();
+            if (!SequenceUtil.isNoCall(readBase)) {
```
My comment is my contribution :P

yfarjoun commented:
@lbergelson I've rebased on top of the latest picard to use the new SamLocusAndReferenceIterator. 

@fleharty I put in a fix for the issue we discussed in private, namely that the Overlapping error was too slow for deep files. I fixed this using a static class which @lbergelson might object to and offer a different solution :-)

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java:
```diff
@@ -0,0 +1,88 @@
+
+    // Updated as a defaulting map.
+    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+    private final CollectionUtil.DefaultingMap<Object, CALCULATOR> strataAggregatorMap = new CollectionUtil.DefaultingMap<>
```
Declared type should be more general, unless a more specific type is required. Using `Map` here should work. Also, the anonymous subclass can be replaced with a lambda.
```
    private final Map<Object, CALCULATOR> strataAggregatorMap = 
            new CollectionUtil.DefaultingMap<>(ignored -> simpleAggregatorGenerator.get(), true);
```

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java:
```diff
@@ -0,0 +1,88 @@
+    public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SAMLocusAndReference locusInfo) {
+        final Object stratus = stratifier.stratify(recordAndOffset, locusInfo);
+        // this assumes we do not want to aggregate null.
+        if (stratus != null) strataAggregatorMap.get(stratus).addBase(recordAndOffset, locusInfo);
```
add `{` `}` and format body across lines

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java:
```diff
@@ -0,0 +1,88 @@
+        final List<ErrorMetrics.ErrorMetricBase> metrics = new ArrayList<>();
+
+        // we do this to sort
+        final TreeSet<Object> strata = new TreeSet<>(strataAggregatorMap.keySet());
```
This is N^2, if you instead create a list and then sort it it would be more efficient. Not sure if that matters here though.

pshapiro4broad replied:
You're right, it's NlogN. I thought add() was N, but it makes sense that it's logN

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java:
```diff
@@ -0,0 +1,88 @@
+ * An interface and implementations for classes that apply a {@link picard.sam.BamErrorMetric.ReadBaseStratification.RecordAndOffsetStratifier RecordAndOffsetStratifier}
+ * to put bases into various "bins" and then compute an {@link ErrorMetrics} on these bases using a {@link BaseErrorCalculation}.
+ *
+ * @author Yossi Farjoun
```
Do we have a policy on `@author` tags? In general I don't think they add much as git blame will tell you who wrote what. Also we never seem to update the tags, so we have author tags for people that no longer work on the code.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java:
```diff
@@ -0,0 +1,88 @@
+        return simpleAggregatorGenerator.get().getSuffix() + "_by_" + stratifier.getSuffix();
+    }
+
+    public ErrorMetrics.ErrorMetricBase[] getMetrics() {
```
I'm not opposed to nested classes but in cases where a class is heavily used outside its parent, it makes for awkward code like this. Could this be moved to a top level type?

Alternatively, you could rename it to `Base` so code that refers to it would now read `ErrorMetrics.Base` which isn't as awkward.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,266 @@
+        long totalBases = 0;
+
+        /**
+         * the function by which new loci are "shown" to the calculator
```
In general, you don't want to javadoc an overridden method. It's unnecessary as the base class javadoc is inherited automatically and, even worse, can lead to skew if the base class doc is changed.

An exception is if the override changes the behavior of a method in a way that a caller would need to know about.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,266 @@
+                    .orElse(null);
+
+            // we are only interested in bases for which the mate read also has a base over the same locus
+            if (mate == null) return;
```
all `if`s should have `{` `}`s

yfarjoun replied:
hmmm. really? I'm not sure that for such simple if statements this style has been adopted and agreed upon.

pshapiro4broad replied:
Well, let's say _IMO_ they should. This is what both the oracle/sun and google java coding guidelines specify. It is also (I believe) the default behavior when using the the intellij java formatter.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculation.java:
```diff
@@ -0,0 +1,266 @@
+            // read2 must be non-secondary
+            if (read2.isSecondaryAlignment()) return false;
+
+            return read1.getMateAlignmentStart() == read2.getAlignmentStart();
```
I think the use of `^` here is unnecessary and makes the code harder to understand. Unless I'm not understanding it. Also I find it's easier to follow as a single `return` but maybe that's just me. e.g.
```
    static private boolean areReadsMates(final SAMRecord read1, final SAMRecord read2) {
        // must have same name
        return read1.getReadName().equals(read2.getReadName()) &&
                // must be paired
                read1.getReadPairedFlag() &&
                // one must be first while the other is not
                !read1.getFirstOfPairFlag() == read2.getFirstOfPairFlag() &&
                // one must be second while the other is not
                !read1.getSecondOfPairFlag() == read2.getSecondOfPairFlag() &&
                // read1 must be mapped
                !read1.getReadUnmappedFlag() &&
                // read2 must be mapped
                !read2.getReadUnmappedFlag() &&
                // read1 must be non-secondary
                !read1.isSecondaryAlignment() &&
                // read2 must be non-secondary
                !read2.isSecondaryAlignment() &&
                read1.getMateAlignmentStart() == read2.getAlignmentStart();
    }
```

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+    @Argument(doc = "Errors to collect in the form of \"ERROR(:STRATIFIER)*\". " +
+            "To see the values available for ERROR and STRATIFIER look at the documentation for the arguments " +
+            "ERROR_VALUE and STRATIFIER_VALUE.")
+    public List<String> ERROR_METRICS = new ArrayList<>(CollectionUtil.makeList(
```
Is there a reason why `Arrays.asList()` isn't used here? It doesn't look like this is modified after its creation.

yfarjoun replied:
It is modified by the magic of the commandline program parsing system.

pshapiro4broad replied:
OK, thought it might be that. In that case, is the `new ArrayList<>()` needed too? It looks like `CollectionUtil.makeList()` returns a modifiable list already.

yfarjoun replied:
true. done.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+            "OVERLAPPING_ERROR:READ_ORDINALITY:GC_CONTENT"));
+
+    @Argument(doc = "A fake argument used to show what the options of ERROR (in ERROR_METRICS) are.", optional = true)
+    public BaseErrorCalculation.Errors ERRORS_VALUE = null;
```
`= null` is unnecessary in Java as that is the default value of an object. same is true of `= 0` below for `long`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+    public int LONG_HOMOPOLYMER = 6;
+
+    @Argument(shortName = "P", doc = "The probability of selecting a locus for analysis (for downsampling)", optional = true, maxValue = 1, minValue = 0)
+    public double PROBABILITY = 1.0d;
```
can just use `= 1;` here

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+        }
+
+        final String[] superValidation = super.customCommandLineValidation();
+        if (superValidation != null) errors.addAll(Arrays.asList(superValidation));
```
need `{` `}`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+        final String[] superValidation = super.customCommandLineValidation();
+        if (superValidation != null) errors.addAll(Arrays.asList(superValidation));
+        if (!errors.isEmpty()) {
+            return errors.toArray(new String[errors.size()]);
```
In general it's better to use `.toArray(new String[0]);`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+        }
+    }
+
+    private final Log log = Log.getInstance(CollectBamErrorMetrics.class);
```
`log` can be static

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+    @Override
+    protected int doWork() {
+
+        final Random random = new Random(42);
```
Is the fixed seed here intentional? Otherwise it looks like `Math.random()` would work here.

yfarjoun replied:
Random seeds are fixed to avoid non-deterministic behavior.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+            }
+
+        } catch (IOException e) {
+            log.error("A problem occurred:", e.getMessage());
```
In general you should always include the exception when logging so a stack crawl is printed. Once you do that you no longer need to get the exception message as it's already included in the output. E.g. `log.error("A problem occurred", e);`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+            aggregatorList.add(aggregator);
+            if (!suffixes.add(aggregator.getSuffix())) {
+                throw new IllegalArgumentException(String.format("Duplicated suffix (%s) found in aggregator %s.",
+                        aggregator.getSuffix(), aggregator.getClass().toString()));
```
unnecessary `.toString()`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+                log.info("Reading IntervalList ", intervalListFile, ".");
+                temp = IntervalList.fromFile(intervalListFile);
+                sequenceDictionary.assertSameDictionary(temp.getHeader().getSequenceDictionary());
+            } else {
```
if you flip the `if` then you can make this just `if (!exists()) { .. throw }` and put everything else at the top level in the `for`. That would avoid having to declare `temp` before it's initialized.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+     * @return The appropriate {@link BaseErrorAggregation} object.
+     */
+    protected static BaseErrorAggregation parseDirective(final String stratifierDirective) {
+        final String[] directiveUnits = new String[256];
```
Magic number `256` duplicated in code, should use a declared constant instead.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+
+        // make a linkedList due to removal and addition operations below
+        final List<ReadBaseStratification.RecordAndOffsetStratifier<?>> stratifiers = new LinkedList<>(Arrays.stream(directiveUnits)
+                .limit(numberOfTerms)
```
Why is `limit` required here if `numberOfTerms` is tested above?

yfarjoun replied:
I don't know...but I get a strange error if I remove it. I'd appreciate help with this.

pshapiro4broad replied:
I just wrote a short test without `limit` and it seemed to work OK. Are you getting a compiler error or a runtime error?

Also when I typed the code into intellij it recommended creating the linked list directly, e.g.
```
        final List<ReadBaseStratification.RecordAndOffsetStratifier<?>> stratifiers =
                Arrays.stream(directiveUnits)
                        .map(String::trim)
                        .map(ReadBaseStratification.Stratifiers::valueOf)
                        .map(ReadBaseStratification.Stratifiers::makeStratifier)
                        .collect(Collectors.toCollection(LinkedList::new));
```

yfarjoun replied:
Yes...I'm getting a strange NPE if I remove that line.

yfarjoun replied:
resolved. this was due to the 256-length array that was allocated.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java:
```diff
@@ -0,0 +1,470 @@
+            while (stratifiers.size() > 1) {
+                ReadBaseStratification.RecordAndOffsetStratifier<?> first = stratifiers.remove(0);
+                ReadBaseStratification.RecordAndOffsetStratifier<?> second = stratifiers.remove(0);
+                ReadBaseStratification.PairStratifier<? extends Comparable, ? extends Comparable> newPair = new ReadBaseStratification.PairStratifier<>(first, second);
```
Maybe it would've been better to have an N-Stratifier in addition to Pair?

yfarjoun replied:
that would have made the signature of the class very confusing (I actually don't know how to do it)

pshapiro4broad replied:
Hmm, the only problem I see is that `Collection` isn't `Comparable`. There's probably a way to wrap this but I don't know it offhand

```
    public static class CollectionStratifier<T extends Comparable<T>> implements RecordAndOffsetStratifier<Collection<T>> {
        final Collection<RecordAndOffsetStratifier<T>> stratifiers;

        public CollectionStratifier(Collection<RecordAndOffsetStratifier<T>> stratifiers) {
            this.stratifiers = stratifiers;
        }

        @Override
        public Collection<T> stratify(RecordAndOffset recordAndOffset, SAMLocusAndReference locusInfo) {
            return stratifiers.stream().map(o -> o.stratify(recordAndOffset, locusInfo)).collect(Collectors.toList());
        }

        @Override
        public String getSuffix() {
            return stratifiers.stream().map(RecordAndOffsetStratifier::getSuffix).collect(Collectors.joining("_and_"));
        }
    }
```

yfarjoun replied:
The problem I was refering to is that in the Pair, left and right can have different types...

pshapiro4broad replied:
> The problem I was refering to is that in the Pair, left and right can have different types

As far as I can tell, in this code fragment the type argument to `Pair` is not bound. This is because the type is generated by a series of calls to `Stratifiers.makeStratifier()`, and the result of that operation is `RecordAndOffsetStratifier<?>`

yfarjoun replied:
I see what you mean. I'll add a CollectionStratifier that will remove this code from the CLP.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetrics.java:
```diff
@@ -0,0 +1,169 @@
+
+        // needed for reading in a metric from a file
+        public OverlappingErrorMetric() {
+            super();
```
`super()` is unnecessary here as it's called automatically if not present.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/Pair.java:
```diff
@@ -0,0 +1,112 @@
+
+        // Check to see whether one is null but not the other.
+        if (this.first == null && other.first != null) return false;
+        if (this.second == null && other.second != null) return false;
```
this can be simplified by using `Objects.equals()`, e.g. `return Objects.equals(first, other.first) && Objects.equals(second, other.second);`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/Pair.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    @Override
+    public int hashCode() {
+        if (second == null && first == null)
```
This can all be replaced with `return Objects.hash(first, second);`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/Pair.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    public String toString() {
+        return first.toString() + "," + second.toString();
```
this is the same as `return first + "," + second;`

yfarjoun commented:
Thanks for the review @pshapiro4broad! 

I feel that the resulting code is much better for it. 

Back to you.

yfarjoun commented:
better now @pshapiro4broad?

pshapiro4broad commented on src/test/java/picard/sam/BamErrorMetric/BaseErrorCalculationTest.java:
```diff
@@ -0,0 +1,153 @@
+        Assert.assertEquals(metric2.NUM_THREE_WAYS_DISAGREEMENT, 1L);
+    }
+
+    @DataProvider( )
```
unnecessary `()`

pshapiro4broad commented on src/test/java/picard/sam/BamErrorMetric/BaseErrorCalculationTest.java:
```diff
@@ -0,0 +1,153 @@
+            builder.getHeader().setSequenceDictionary(referenceSequenceFileWalker.getSequenceDictionary());
+
+            for (int i = 0; i < 4000; i++) {
+                builder.addPair("Read" + String.valueOf(i), 0, 1, 1,
```
unnecessary `String.valueOf()`

pshapiro4broad commented on src/test/java/picard/sam/BamErrorMetric/BaseErrorCalculationTest.java:
```diff
@@ -0,0 +1,153 @@
+                builder.forEach(writer::addAlignment);
+            }
+        }
+        return new Object[][]{{temp}};
```
Doesn't seem to be a benefit to having a dataprovider that has a single return value and only one use. Making this a method like `createTestDataFileForXXXX()` and having it return `File` might be clearer.

yfarjoun replied:
I wanted to separate the time for creation of the file from the time or the test....

pshapiro4broad commented on src/test/java/picard/sam/BamErrorMetric/BaseErrorAggregationTest.java:
```diff
@@ -0,0 +1,192 @@
+import java.util.List;
+
+/**
+ * Created by farjoun on 5/31/18.
```
Like `@author` tags, a created by comment doesn't seem very helpful to me.

pshapiro4broad commented on src/test/java/picard/sam/BamErrorMetric/BaseErrorAggregationTest.java:
```diff
@@ -0,0 +1,192 @@
+        BaseErrorAggregation<?> baseErrorAggregation = new BaseErrorAggregation<>(OverlappingReadsErrorCalculator::new, ReadBaseStratification.readOrdinalityStratifier);
+        final int length = refBases.length;
+
+        for (int i = 0; i < length; i++) {
```
It looks like this `for` loop pattern occurs many times in different tests. It would be better to replace the duplicate code with a common method.

pshapiro4broad commented on src/test/java/picard/sam/BamErrorMetric/BaseErrorAggregationTest.java:
```diff
@@ -0,0 +1,192 @@
+
+        final ErrorMetric[] metrics = baseErrorAggregation.getMetrics();
+        final BaseErrorMetric metricA = Arrays.stream(metrics).map(a -> (BaseErrorMetric) a)
+                .filter(a -> a.COVARIATE.equals("A")).findFirst().orElse(null);
```
The code calls `orElse(null)` and on the next line makes a method call, generating a NPE. This is OK for a test but IMO using `.get()` would a simpler way to write it.

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/SimpleErrorCalculator.java:
```diff
@@ -0,0 +1,65 @@
+ */
+public class SimpleErrorCalculator extends BaseErrorCalculator {
+
+    protected long nMismatchingBases = 0;
```
unnecessary initializer

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/OverlappingReadsErrorCalculator.java:
```diff
@@ -0,0 +1,146 @@
+ * but the other does not which indicates that there might have been a sequencing error in that read.
+ */
+public class OverlappingReadsErrorCalculator extends BaseErrorCalculator {
+    private long nBothDisagreeWithReference = 0;
```
unnecessary initializers

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/ErrorMetric.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class ErrorMetric extends MergeableMetricBase {
+
+    public static void setPriorError(double priorError) {
```
Not sure what the picard convention is, but it's typical in Java code to sort fields before methods

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorMetric.java:
```diff
@@ -0,0 +1,28 @@
+
+    /** The (phred) rate of bases that disagree with the reference */
+    @NoMergingIsDerived
+    public Integer Q_SCORE;
```
use `int` unless `null` is allowed here

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculator.java:
```diff
@@ -0,0 +1,44 @@
+import htsjdk.samtools.util.SequenceUtil;
+
+public abstract class BaseErrorCalculator implements BaseCalculator {
+    long totalBases = 0;
```
unnecessary initializer.

Also, the code seems to use both `nSomething` and `totalSomething`. It seems like it would be better to use a single convention in the code. (compare with `SimpleErrorCalculator.nMismatchingBases`)

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/BaseErrorCalculator.java:
```diff
@@ -0,0 +1,44 @@
+     **/
+    @Override
+    public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SamLocusAndReferenceIterator.SAMLocusAndReference locusAndRef) {
+        final byte readBase = recordAndOffset.getReadBase();
```
unnecessary local variable

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,916 @@
+            // not including the current base.
+            if (recordAndOffset.getReadBase() != locusInfo.getReferenceBase()) {
+                return numberMismatches - 1;
+            } else {
```
`else` unnecessary after `return`

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,916 @@
+
+        @Override
+        public Integer stratify(final SAMRecord sam) {
+            final int tile;
```
unnecessary local variable, code can just `return location.getTile();` below

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,916 @@
+                tile = location.getTile();
+                return tile;
+            } catch (final IllegalArgumentException ignored) {
+                return null;
```
Would there be a benefit to logging the caught exception here?

pshapiro4broad commented on src/main/java/picard/sam/BamErrorMetric/ReadBaseStratification.java:
```diff
@@ -0,0 +1,916 @@
+     * <p>
+     * This is used in {@link CollectBamErrorMetrics} to convert an input argument to a fully functional {@link BaseErrorAggregation} object.
+     */
+    enum Stratifiers implements CommandLineParser.ClpEnum {
```
In general enum names should be singular since it's not a collection of things, but a type. E.g. if you have one stratifier in your code you'd want declare it as `Stratifier s;` not `Stratifiers s;`

yfarjoun commented on src/main/java/picard/sam/BamErrorMetric/Pair.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public int compareTo(final Pair<X, Y> o) {
+        if (this.left == null || o.left == null)
```
hmmm. So in brave world what is Int(5).compareTo(null)?

yfarjoun replied:
Oh I see, you're saying null==null so compare of null with null should return 0....ok....edge case, but sure.

yfarjoun commented on src/test/java/picard/sam/BamErrorMetric/CollectBamErrorMetricsTest.java:
```diff
@@ -0,0 +1,328 @@
+import java.util.Set;
+
+public class CollectBamErrorMetricsTest {
+    private static final File OUTPUT_DATA_PATH = IOUtil.createTempDir("CollectBamErrorMetricsTest", null);
```
good call!

yfarjoun commented on src/test/java/picard/sam/BamErrorMetric/CollectBamErrorMetricsTest.java:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    @Test
+    void testAggregatorsHaveDistinctSuffixes() throws InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
```
yeah...stagnant code...sorry 'bout that.

yfarjoun commented:
thanks for all the comments @pshapiro4broad. If I understand your last comment correctly, I can merge now.....if I misunderstood, please let me know and I'll submit a correction PR.

yfarjoun commented:
🎉
